### PR TITLE
Make sure the absolute path isn't contained in the cache

### PR DESCRIPTION
**Pull request by mischnic**
Depends on https://github.com/parcel-bundler/parcel/pull/5802

Closes T-870

This should also always use `/` as path separators

Grepping the absolute path in .parcel-cache didn't yield any results after this change (for a simple test case).

- [x] I wasn't yet able to figure out why the the remaining few tests fail. Something with cache invalidation after file changes is wrong.
- [x] Move the new utils into `utils.js` in `@parcel/core`, no reason to put them into `@parcel/utils` if they're not used outside of core.
- [x] locs
- [ ] still use micromatch 4?
- [x] throw an error if namers return an absolute path
- [x] Windows CI fails?

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 8.99s | -202.00ms |
| Cached | 447.00ms | +20.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/logo.1e014c76.png | 274.00b | +0.00b | 203.00ms | -63.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/logo.1e014c76.png | 274.00b | +0.00b | 225.00ms | -75.00ms 🚀 |

</p></details>

<details><summary>AtlasKit Editor 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 6.74s | -151.00ms |
| Cached | 431.00ms | +39.00ms ⚠️ |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/303424942251180609)**

File: packages/core/core/test/TargetRequest.test.js [hunk 1]
```diff
@@ -179,6 +179,7 @@ describe('TargetResolver', () => {
             sourceMap: {},
           },
           loc: {
+            // TODO?
             filePath: path.join(COMMON_TARGETS_FIXTURE_PATH, 'package.json'),
             start: {
```
Comment (mischnic) [inline packages/core/core/test/TargetRequest.test.js]: I guess the locs should also be relative to the root

File: packages/core/core/src/types.js [hunk 2]
```diff
@@ -46,6 +45,7 @@ export type ExtendableParcelConfigPipeline = $ReadOnlyArray<
 >;
 
 export type ProcessedParcelConfig = {|
+  // TODO ProjectPath?
   extends?: PackageName | FilePath | Array<PackageName | FilePath>,
```
Comment (mischnic) [inline packages/core/core/src/types.js]: Haven't determined yet if this is ever an absolute path (and if it's stored in the cache at all)

File: packages/core/core/src/projectPath.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+export opaque type ProjectPath = string;
+
+export function toProjectPath_(
+  projectRoot: FilePath,
+  p: FilePath,
+): ProjectPath {
+  // TODO we could save the `./` at the start of every path
+  // return p != null ? relativePath(projectRoot, p, false) : p;
+  return p != null ? relativePath(projectRoot, p) : p;
```
Comment (mischnic) [inline packages/core/core/src/projectPath.js]: ...

File: packages/core/core/src/RequestTracker.js [hunk 4]
```diff
@@ -572,15 +592,15 @@ export class RequestGraph extends Graph<
...
+            this.invalidateFileNameNode(fileNameNode, _filePath, above);
           }
         }
 
         for (let id of this.globNodeIds) {
           let globNode = this.getNode(id);
           invariant(globNode && globNode.type === 'glob');
 
-          if (isGlobMatch(filePath, globNode.value)) {
+          if (isGlobMatch(filePath, fromProjectPathRelative(globNode.value))) {
```
Comment (devongovett) [inline packages/core/core/src/RequestTracker.js:618]: So `filePath` here is absolute, but the glob is relative? Won't that not match except if the glob starts with `**/`? I was thinking about how to do this for the glob resolver. Should the glob in the invalidation be relative to the project root? If so, then I think we'd need to make it absolute here.

Comment (mischnic) [reply packages/core/core/src/RequestTracker.js:618]: `filePath` is also relative here, further up there's

```js
let filePath = fromProjectPathRelative(_filePath);
```

Comment (mischnic) [reply packages/core/core/src/RequestTracker.js:618]: But I ran into https://github.com/parcel-bundler/parcel/pull/5483#issuecomment-746479825 again, so I guess I'll have to use the absolute paths anyway (to make micromatch happy).

As a workaround, I've downgraded to micromatch 3 for now 😬

File: packages/core/core/src/projectPath.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+export const toProjectPath: ((
+  projectRoot: FilePath,
+  p: FilePath,
+) => ProjectPath) &
+  ((projectRoot: FilePath, p: FilePath | void) => ProjectPath | void) &
+  // $FlowFixMe Not sure how to type properly
+  ((projectRoot: FilePath, p: ?FilePath) => ?ProjectPath) = toProjectPath_;
```
Comment (mischnic) [inline packages/core/core/src/projectPath.js:30]: 🤷‍♂️

File: packages/core/core/src/types.js (hunk 2 above)
Comment (mischnic) [reply packages/core/core/src/types.js]: This property was apparently unused, so I just removed it...

**Comment by elliottkember** [comment]
Hi! Great work here. this PR would be a really big deal for CI / deployment environments where the build directory changes between builds. Several orders of magnitude difference. At the moment it seems the Parcel cache uses the absolute directory.

File: packages/core/types/index.js [hunk 6]
```diff
@@ -445,10 +445,9 @@ export type DependencyOptions = {|
   +meta?: Meta,
   +pipeline?: string,
   +resolveFrom?: FilePath,
-  +target?: Target,
```
Comment (mischnic) [inline packages/core/types/index.js]: This was apparently neither used in the plugins nor inside `addDependency`

**Comment by devongovett** [comment]
One interesting problem is what to do about source maps. The source map object stores the project root at the moment, and source map buffers are stored in the cache. Perhaps we can make the project root a required parameter of the sourcemap constructor even if you are creating it from a buffer, and not include that as part of the serialized payload? @DeMoorJasper

**Comment by devongovett** [comment]
Unfortunately, all of the path conversion seems to add around ~300ms to the esbuild benchmark. We might need to find a faster way of doing this... 😞

**Comment by devongovett** [comment]
Did some optimizations and now the perf is pretty much the same as v2.

**Comment by devongovett** [comment]
Also now storing paths that are outside the project root as absolute rather than relative since they are not portable anyway. This means that at least if the project root moves, the file references with still work (unless the referenced files also moved). Thoughts?

**Comment by elliottkember** [comment]
> Also now storing paths that are outside the project root as absolute rather than relative since they are not portable anyway. This means that at least if the project root moves, the file references with still work (unless the referenced files also moved). Thoughts?

Sounds fine to me!

**Comment by devongovett** [comment]
I think this is ready. I added a util that verifies that there are no references to the project root inside the cache folder after each cache test. After upgrading the source maps package to the latest release, all the tests pass.

File: packages/core/integration-tests/test/cache.js [hunk 7]
```diff
@@ -69,13 +75,33 @@ async function testCache(update: UpdateFn | TestConfig, integration) {
     }
   }
 
+  let resolvedOptions = await resolveOptions(
+    getParcelOptions(getEntries(entries), getOptions(options)),
+  );
+
   let b = await runBundle(entries, options);
 
+  await assertNoFilePathInCache(
```
Comment (wbinnssmith) [inline packages/core/integration-tests/test/cache.js:84]: 🤩

**Review by wbinnssmith — APPROVED** [review]
It's a lot, but from what I can tell LGTM. cc @mischnic to review changes from @devongovett

**Pull request merged**